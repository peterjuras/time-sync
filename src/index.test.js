import TimeSync from './index';
import lolex from 'lolex';

describe('#TimeSync', () => {
  const instance = new TimeSync();
  let clock;

  beforeEach(() => {
    clock = lolex.install({
      now: 1
    });
  });

  afterEach(() => {
    clock.uninstall();
    instance.removeAllTimers();
  });

  it('should be exported correctly', () => {
    expect(instance).toBeInstanceOf(TimeSync);
  });

  describe('#constants', () => {
    it('SECONDS should be exported correctly', () => {
      expect(TimeSync.SECONDS).toBe('s');
    });

    it('MINUTES should be exported correctly', () => {
      expect(TimeSync.MINUTES).toBe('m');
    });

    it('HOURS should be exported correctly', () => {
      expect(TimeSync.HOURS).toBe('h');
    });

    it('DAYS should be exported correctly', () => {
      expect(TimeSync.DAYS).toBe('d');
    });
  });

  describe('#addTimer', () => {
    it('should be exported correctly', () => {
      expect(instance.addTimer).toBeInstanceOf(Function);
    });

    it('should throw if no callback is provided', () => {
      expect(instance.addTimer).toThrowErrorMatchingSnapshot();
    });

    it('should throw if for an unknown interval', () => {
      expect(() => instance.addTimer(jest.fn(), {
        interval: []
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        interval: {}
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        interval: 123
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        interval: ''
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        interval: 'a'
      })).toThrowErrorMatchingSnapshot();
    });

    it('should throw if for an invalid unit', () => {
      expect(() => instance.addTimer(jest.fn(), {
        unit: []
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: {}
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: ''
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: 'a'
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: 250.291
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: 0
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: -12
      })).toThrowErrorMatchingSnapshot();

      expect(() => instance.addTimer(jest.fn(), {
        unit: -12.28
      })).toThrowErrorMatchingSnapshot();
    });

    it('should add a second timer by default', () => {
      const mock = jest.fn();
      instance.addTimer(mock);

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(998);
      expect(mock).toHaveBeenCalledTimes(0);

      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(1);

      clock.tick(999);
      expect(mock).toHaveBeenCalledTimes(1);
    });

    it('should correctly callback for a second timer', () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.SECONDS });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick((1000 * 2) - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(1000);
    });

    it('should correctly callback for a minute timer', () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.MINUTES });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick((60000 * 2) - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(60000);
    });

    it('should correctly callback for a hour timer', () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.HOURS });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick((1000 * 60 * 60 * 2) - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(1000 * 60 * 60);
    });

    it('should correctly callback for a day timer', () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.DAYS });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick((1000 * 60 * 60 * 24 * 2) - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(1000 * 60 * 60 * 24);
    });

    it('should work with a mix of timers', () => {
      const mockSecond = jest.fn();
      const mockTenSeconds = jest.fn();
      const mockMinute = jest.fn();
      instance.addTimer(mockSecond);
      instance.addTimer(mockTenSeconds, { unit: 10 });
      instance.addTimer(mockMinute, { interval: TimeSync.MINUTES });

      clock.tick((1000 * 60) - 1);
      expect(mockSecond).toHaveBeenCalledTimes(60);
      expect(mockTenSeconds).toHaveBeenCalledTimes(6);
      expect(mockMinute).toHaveBeenCalledTimes(1);
    });

    it('should clean up timeouts if the last timer is removed', () => {
      const mock = jest.fn();
      const ref = instance.addTimer(mock);

      clock.tick(1000);
      expect(mock).toHaveBeenCalledTimes(1);
      ref();

      clock.tick(1000);
      expect(mock).toHaveBeenCalledTimes(1);
    });
  });

  describe('#getLastTime', () => {
    it('should be exported correctly', () => {
      expect(instance.getLastTime).toBeInstanceOf(Function);
    });

    it('should return the correct time if it has not been initialized yet', () => {
      const now = Date.now();

      expect(instance.getLastTime()).toBe(now);

      clock.tick(50);

      expect(instance.getLastTime()).toBe(now);
    });

    it('should return the correct time if it has been advanced by a timer', () => {
      const testTime = Date.now();
      expect(instance.getLastTime()).toBe(testTime);
      clock.tick(100);

      const mock = jest.fn();
      instance.addTimer(mock);
      clock.tick(1000);

      expect(instance.getLastTime()).toBe(1000);
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(instance.getLastTime());
    });

    it('should correctly remove timer if it schedules for the next tick', () => {
      const mock1 = jest.fn();
      const mock2 = jest.fn();
      const mock3 = jest.fn();
      const mock4 = jest.fn();
      instance.addTimer(mock1);
      const removeReference = instance.addTimer(mock2);
      instance.addTimer(mock3, { interval: TimeSync.MINUTES });
      instance.addTimer(mock4);

      // Advance to just before two minute
      clock.tick((1000 * 60) - 2);

      expect(mock1).toHaveBeenCalledTimes(59);
      expect(mock2).toHaveBeenCalledTimes(59);
      expect(mock3).toHaveBeenCalledTimes(0);
      expect(mock4).toHaveBeenCalledTimes(59);

      removeReference();

      clock.tick(200);

      expect(mock1).toHaveBeenCalledTimes(60);
      expect(mock2).toHaveBeenCalledTimes(59);
      expect(mock3).toHaveBeenCalledTimes(1);
      expect(mock4).toHaveBeenCalledTimes(60);
    });

    it('removing a timer should not throw if it is called after all timers have been removed', () => {
      const mock = jest.fn();
      const ref = instance.addTimer(mock);

      instance.removeAllTimers();
      ref();
    });
  });

  describe('#removeAllTimers', () => {
    it('should be exported correctly', () => {
      expect(instance.removeAllTimers).toBeInstanceOf(Function);
    });
  });

  describe('#revalidate', () => {
    it('should be exported correctly', () => {
      expect(instance.revalidate).toBeInstanceOf(Function);
    });

    it('should fire all necessary timers if system clock has changed', () => {
      const mockSecond = jest.fn();
      const mockTenSeconds = jest.fn();
      const mockMinute = jest.fn();
      instance.addTimer(mockSecond);
      instance.addTimer(mockTenSeconds, { unit: 10 });
      instance.addTimer(mockMinute, { interval: TimeSync.MINUTES });

      clock.setSystemTime(1000 * 60 * 60);

      expect(mockSecond).toHaveBeenCalledTimes(0);
      expect(mockTenSeconds).toHaveBeenCalledTimes(0);
      expect(mockMinute).toHaveBeenCalledTimes(0);

      clock.tick();

      expect(mockSecond).toHaveBeenCalledTimes(0);
      expect(mockTenSeconds).toHaveBeenCalledTimes(0);
      expect(mockMinute).toHaveBeenCalledTimes(0);

      instance.revalidate();
      clock.tick();

      expect(mockSecond).toHaveBeenCalledTimes(1);
      expect(mockTenSeconds).toHaveBeenCalledTimes(1);
      expect(mockMinute).toHaveBeenCalledTimes(1);
    });

    it('should revalidate added timers', () => {
      const mock1 = jest.fn();
      const mock2 = jest.fn();
      instance.addTimer(mock1);

      clock.setSystemTime(1250);
      expect(mock1).toHaveBeenCalledTimes(0);
      expect(mock2).toHaveBeenCalledTimes(0);

      instance.addTimer(mock2);
      expect(mock1).toHaveBeenCalledTimes(0);
      expect(mock2).toHaveBeenCalledTimes(0);

      instance.revalidate();
      clock.tick(1);
      expect(mock1).toHaveBeenCalledTimes(1);
      expect(mock2).toHaveBeenCalledTimes(1);

      clock.tick(749);
      expect(mock1).toHaveBeenCalledTimes(2);
      expect(mock2).toHaveBeenCalledTimes(2);
    });

    it('should revalidate even if next timer has not fired yet', () => {
      const mock = jest.fn();
      instance.addTimer(mock);

      clock.setSystemTime(250);
      instance.revalidate();

      clock.tick(750);
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(1000);
    });

    it('should resync automatically after any timer is fired', () => {
      const mock1 = jest.fn();
      const mock2 = jest.fn();
      const mock3 = jest.fn();
      instance.addTimer(mock1);
      instance.addTimer(mock2, { unit: 2 });
      instance.addTimer(mock3, { unit: 2 });

      clock.setSystemTime(2500);
      clock.tick();

      expect(mock1).toHaveBeenCalledTimes(1);
      expect(mock2).toHaveBeenCalledTimes(1);
      expect(mock3).toHaveBeenCalledTimes(1);
    });
  });
});
