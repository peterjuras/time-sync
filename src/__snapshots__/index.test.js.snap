// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#TimeSync #addTimer should throw if for an invalid unit 1`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 2`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 3`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 4`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 5`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 6`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 7`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 8`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an invalid unit 9`] = `"unit must be a positive integer"`;

exports[`#TimeSync #addTimer should throw if for an unknown interval 1`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#TimeSync #addTimer should throw if for an unknown interval 2`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#TimeSync #addTimer should throw if for an unknown interval 3`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#TimeSync #addTimer should throw if for an unknown interval 4`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#TimeSync #addTimer should throw if for an unknown interval 5`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#TimeSync #addTimer should throw if no callback is provided 1`] = `"You need to provide a callback as the first argument to addTimer"`;

exports[`#TimeSync #addTimer should work with a mix of timers 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
  Array [
    4,
  ],
  Array [
    5,
  ],
  Array [
    6,
  ],
  Array [
    7,
  ],
  Array [
    8,
  ],
  Array [
    9,
  ],
  Array [
    10,
  ],
  Array [
    11,
  ],
  Array [
    12,
  ],
  Array [
    13,
  ],
  Array [
    14,
  ],
  Array [
    15,
  ],
  Array [
    16,
  ],
  Array [
    17,
  ],
  Array [
    18,
  ],
  Array [
    19,
  ],
  Array [
    20,
  ],
  Array [
    21,
  ],
  Array [
    22,
  ],
  Array [
    23,
  ],
  Array [
    24,
  ],
  Array [
    25,
  ],
  Array [
    26,
  ],
  Array [
    27,
  ],
  Array [
    28,
  ],
  Array [
    29,
  ],
  Array [
    30,
  ],
  Array [
    31,
  ],
  Array [
    32,
  ],
  Array [
    33,
  ],
  Array [
    34,
  ],
  Array [
    35,
  ],
  Array [
    36,
  ],
  Array [
    37,
  ],
  Array [
    38,
  ],
  Array [
    39,
  ],
  Array [
    40,
  ],
  Array [
    41,
  ],
  Array [
    42,
  ],
  Array [
    43,
  ],
  Array [
    44,
  ],
  Array [
    45,
  ],
  Array [
    46,
  ],
  Array [
    47,
  ],
  Array [
    48,
  ],
  Array [
    49,
  ],
  Array [
    50,
  ],
  Array [
    51,
  ],
  Array [
    52,
  ],
  Array [
    53,
  ],
  Array [
    54,
  ],
  Array [
    55,
  ],
  Array [
    56,
  ],
  Array [
    57,
  ],
  Array [
    58,
  ],
  Array [
    59,
  ],
  Array [
    60,
  ],
]
`;

exports[`#TimeSync #addTimer should work with a mix of timers 2`] = `
Array [
  Array [
    10,
  ],
  Array [
    20,
  ],
  Array [
    30,
  ],
  Array [
    40,
  ],
  Array [
    50,
  ],
  Array [
    60,
  ],
]
`;

exports[`#TimeSync #addTimer should work with a mix of timers 3`] = `
Array [
  Array [
    60,
  ],
]
`;
